src/Mapbender/CoreBundle/Resources/public/mapbender.element.legend.js

```diff
@@ -1,515 +1,478 @@
+            displayType:              "list",
+            checkGraphic:             false,
+            hideEmptyLayers:          true,
             generateLegendGraphicUrl: false,
-            showSourceTitle: true,
-            showLayerTitle: true,
-            showGrouppedTitle: true,
-            maxImgWidth: 0,
-            maxImgHeight: 0
+            showSourceTitle:          true,
+            showLayerTitle:           true,
+            showGroupedTitle:         true,
+            maxImgWidth:              0,
+            maxImgHeight:             0
         },
-        model: null,
-        layerTitle: "",
-        sourceTitle: "",
-        grouppedTitle: "",
-        hiddeEmpty: "",
-        _create: function(){
-            if(!Mapbender.checkTarget("mbLegend", this.options.target)) {
+
+        readyCallbacks: [],
+        callback:       null,
+        model:          null,
+
+        /**
+         * Widget constructor
+         *
+         * @private
+         */
+        _create: function() {
+            var widget = this;
+            var options = widget.options;
+            if(!Mapbender.checkTarget("mbLegend", options.target)) {
                 return;
             }
-            var self = this;
-            Mapbender.elementRegistry.onElementReady(this.options.target, $.proxy(self._setup, self));
+            Mapbender.elementRegistry.onElementReady(options.target, $.proxy(widget._setup, widget));
         },
-        _setup: function(){
-            var self = this;
-            this.options.noLegend = Mapbender.trans("mb.core.legend.nolegend");
-            this.model = $("#" + self.options.target).data("mapbenderMbMap").getModel();
-
-            this.layerTitle = this.options.showLayerTitle ? "" : "notshow";
-            this.sourceTitle = this.options.showSourceTitle ? "" : "notshow";
-            this.grouppedTitle = this.options.showGrouppedTitle ? "" : "notshow";
-            this.hiddeEmpty = this.options.hideEmptyLayers ? "notshow" : "";
-
-            if(this.options.autoOpen) {
-                this.open();
-            } else if(this.options.elementType !== 'dialog') {
-                var self = this;
-                var sources = this._getSources();
-                if(this.options.checkGraphic) {
-                    this._createCheckedLegendHtml(sources, 0, 0, "", "", $.proxy(self._createLegend, self));
-                } else {
-                    this._createLegend(this._createLegendHtml(sources));
-                }
-                sources = this.model.getSources();
-                for(var i = 0; i < sources.length; i++) {
-                    this._checkLayers(sources[i]);
-                }
+
+        /**
+         * Setup widget
+         *
+         * @private
+         */
+        _setup: function() {
+            var widget = this;
+            var options = widget.options;
+            var mapWidget = $("#" + widget.options.target).data("mapbenderMbMap");
```

Comment:
I'm not happy to search the target by an ID string instead of using the HTMLElement directly, but I think this is a deeper change to the architecture.

But we should keep in mind, that we should refactor this.

Reply:
@LazerTiberius Check, if the map is available? Scoping?